1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,1322 @@
+except they don't set `$LASTEXITCODE`
+and have other ways of expressing error conditions.
+
+Here the same principle applies as with native commands;
+the statements proceed as if the next is
+wrapped in `if ($?) { ... }`.
+
+```powershell
+Write-Output "Hello" && Write-Output "Hello again"
+```
+
+```output
+Hello
+Hello again
+```
+
+---
+
+```powershell
+Write-Output "Hello" || Write-Output "Hello again"
+```
+
+```output
+Hello
+```
+
+---
+
+```powershell
+Write-Error "Bad" && Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+```
+
+`Write-Error` here emits a non-terminating error
+and so we proceed to evaluate `$?`.
+
+---
+
+```powershell
+Write-Error "Bad" || Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+
+Hello again
+```
+
+```powershell
+echo 'Hi' && Write-Output 'Hello'
+```
+
+```output
+Hi
+Hello
+```
+
+---
+
+```powershell
+error 'Bad' && Write-Output 'Hello'
+```
+
+```
+Bad
+```
+
+---
+
+```powershell
+Write-Error 'Bad' || echo 'Message'
+```
+
+```output
+Bad
+Message
+```
+
+---
+
+### Pipelines
+
+Pipeline chains allow whole pipelines between chain operators.
+As above, when a pipeline ends other than with a terminating error,
+`$?` determines the chain logic.
+
+The whole pipeline on the left-hand side of an operator
+will be evaluated before evaluating chain condition
+and then right-hand side.
+
+```powershell
+1,2,3 | ForEach-Object { $_ + 1 } && Write-Output 'Hello'
+```
+
+```output
+2
+3
+4
+Hello
+```
+
+---
+
+```powershell
+1,2,3 | ForEach-Object { if ($_ -eq 2) { Write-Error 'Bad' } else { $_ } } && Write-Output 'Hello'
+```
+
+```output
+1
+1,2,3 | ForEach-Object { if ($_ -eq 2) { Write-Error 'Bad' } else { $_ } } && Write-Output 'Hello' : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+
+3
+Hello
+```
+
+---
+
+```powershell
+function FailInProcess
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(ValueFromPipeline)]
+        $Value
+    )
+
+    process
+    {
+        if ($_ -eq 3)
+        {
+            $err = Write-Error 'Bad' 2>&1
+            $PSCmdlet.WriteError($err)
+            return
+        }
+
+        $PSCmdlet.WriteObject($_)
+    }
+}
+
+1,2,3,4 | FailInProcess && Write-Output 'Succeeded'
+```
+
+```output
+1
+2
+FailInProcess : Bad
+At line:22 char:11
++ 1,2,3,4 | FailInProcess && Write-Output 'Succeeded'
++           ~~~~~~~~~~~~~
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,FailInProcess
+
+4
+```

```

Comment:
FYI, the inconsistency between this example not succeeding and the previous example succeeding is the crux of a discussion within @PowerShell/powershell-committee. 

The real question is whether we want to fix `$?` before or after relying upon it in these chain operators.

Reply:
I’d rather we fix `$?` before relying on it here.

Reply:
@PowerShell/powershell-committee agrees we shouldn't block this functionality on fixing `$?`. It's useful today, and this new functionality will be useful in spite of the quirks. 

In my opinion, it's also going to help drag issues with `$?` out into the light.

Reply:
@joeyaiello should we file an issue around the problematic nature of `$?` to track that?

Reply:
That'd be awesome, though I suspect we have a very old one (2017 or earlier) on error handling in general where a lot of this is already captured.

Reply:
We have https://github.com/PowerShell/PowerShell/issues/3359, though it comes with a lot of discussion - if we now agree that it should be fixed, perhaps opening a new, focused one is called for.